runtime/commonMain/src/kotlinx/serialization/KSerializer.kt

```diff
@@ -0,0 +1,179 @@
+ * encoding/decoding is abstracted from a particular type and its serial form and is responsible
+ * for transforming primitives ("here in an int property 'foo'" call from a serializer) into a particular
+ * format-specific representation ("for a given int, append a property name in quotation marks,
+ * then append a semicolon, then append an actual value" for JSON) and how to retrieve a primitive
+ * ("give me an int that is 'foo' property") from the underlying representation ("expect the next string to be 'foo',
+ * parse it, then parse semicolon, then parse a string until the next comma as an int an return it).
+ *
+ * Serial form consists of a structural description, declared by the [descriptor] and
+ * actual serialization and deserialization processes, defined by the corresponding
+ * [serialize] and [deserialize] methods implementation.
+ *
+ * Structural description specifies how the [T] is represented in the serial form:
+ * its [kind][SerialKind] (e.g. whether it is represented as a primitive, a list or a class),
+ * its [elements][SerialDescriptor.elementDescriptors] and their [positional names][SerialDescriptor.getElementName].
+ *
+ * Serialization process is defined as a sequence of calls to an [Encoder], and transforms a type [T]
+ * into a stream of format-agnostic primitives that represent [T], such as "here is an int, here is a double
+ * and here is another nested object". It can be demonstrated by the example:
+ * ```
+ * class MyData(int: Int, stringList: List<String>, alwaysZero: Long)
+ *
+ * // .. serialize method of a corresponding serializer
+ * fun serialize(encoder: Encoder, value: MyData): Unit = encoder.encodeStructure(descriptor) {
+ *     // encodeStructure encodes beginning and end of the structure
+ *     // encode 'int' property as Int
+ *     encodeIntElement(descriptor, index = 0, value.int)
+ *     // encode 'stringList' property as List<String>
+ *     encodeSerializableElement(descriptor, index = 1, serializer<List<String>>, value.stringList)
+ *     // don't encode 'alwaysZero' property because we decided to do so
+ * } // end of the structure
+ * ```
+ *
+ * Deserialization process is symmetric and uses [Decoder].
+ */
+public interface KSerializer<T> : SerializationStrategy<T>, DeserializationStrategy<T> {
+    /**
+     * Describes the structure of the serializable representation of [T], produced
+     * by this serializer. Knowing the structure of the descriptor is required to determine
+     * the shape of the serialized form (e.g. what elements are encoded as lists and what as primitives)
+     * along with its metadata such as alternative names.
+     *
+     * The descriptor is used dynamically, during serialization by encoders and decoders
+     * to introspect the type and metadata of [T]'s elements being encoded or decoded and
+     * statically, to introspect the type, infer the schema or to compare against the predefined schema.
```

Comment:
I'm not sure 'dynamic/static' is a good wording. What's dynamic in encoders and what's static in schema generators?